codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java

```diff
@@ -0,0 +1,373 @@
+    public static final String NO_DOCUMENT_SERDE_TYPE_NAME = "noSmithyDocumentSerde";
+    public static final String NEW_LAZY_DOCUMENT = "NewLazyDocument";
+    public static final String INTERNAL_NEW_DOCUMENT_MARSHALER_FUNC = "NewDocumentMarshaler";
+    public static final String INTERNAL_NEW_DOCUMENT_UNMARSHALER_FUNC = "NewDocumentUnmarshaler";
+    public static final String INTERNAL_IS_DOCUMENT_INTERFACE = "IsInterface";
+    public static final String UNMARSHAL_SMITHY_DOCUMENT_METHOD = "UnmarshalSmithyDocument";
+    public static final String MARSHAL_SMITHY_DOCUMENT_METHOD = "MarshalSmithyDocument";
+
+    private static final String SERVICE_SMITHY_DOCUMENT_INTERFACE = "smithyDocument";
+    private static final String IS_SMITHY_DOCUMENT_METHOD = "isSmithyDocument";
+
+
+    private final GoSettings settings;
+    private final GoDelegator delegator;
+    private final Set<DocumentShape> documentShapes = new TreeSet<>();
+    private final Model model;
+
+    public ProtocolDocumentGenerator(
+            GoSettings settings,
+            Model model,
+            GoDelegator delegator
+    ) {
+        this.settings = settings;
+        this.model = model;
+        this.delegator = delegator;
+    }
+
+    /**
+     * Get the set of document shapes for the service.
+     *
+     * @return the service's document shapes.
+     */
+    public Set<DocumentShape> getDocumentShapes() {
+        return documentShapes;
+    }
+
+    /**
+     * Returns whether the service has any document shapes.
+     *
+     * @return whether the service has one or more document types.
+     */
+    public boolean hasDocumentShapes() {
+        return getDocumentShapes().size() > 0;
+    }
+
+    /**
+     * Add a document shape to the generator.
+     *
+     * @param shape the document shape to add.
+     */
+    public void addDocumentShape(DocumentShape shape) {
+        documentShapes.add(shape);
+    }
+
+    /**
+     *
+     */
+    public void generateStandardTypes() {
+        generateNoSerdeType();
+        generateInternalDocumentInterface();
+        generateDocumentPackage();
+    }
+
+    private void generateDocumentPackage() {
+        if (!hasDocumentShapes()) {
+            return;
+        }
+
+        writeDocumentPackage("document.go", writer -> {
+            writer.write("type $L = $T", DOCUMENT_INTERFACE_NAME,
+                            getInternalDocumentSymbol(DOCUMENT_INTERFACE_NAME))
+                    .write("");
+
+            writer.openBlock("func $L(v interface{}) $T {", "}", NEW_LAZY_DOCUMENT,
+                            getDocumentSymbol(DOCUMENT_INTERFACE_NAME), () -> {
+                                writer.write("return $T(v)",
+                                        getInternalDocumentSymbol(INTERNAL_NEW_DOCUMENT_MARSHALER_FUNC));
+                            })
+                    .write("");
+        });
+    }
+
+    private void generateNoSerdeType() {
```

Comment:
This type is actually needed in all service client structures types to avoid making backwards incompatible changes when modeled smithy document types are later added at a future date. Since Smithy makes no opinionated stance currently on things like JSON object keys should be formatted (lower-case, camelCase, UPPER CASE) etc, this NoSerde type is embedded in a generated smithy structure types from the model to prevent their usage as document types. That way if an existing structure is later annotated as modeled document type, we don't have a situation where we make a potential breaking change in the representation of the underlying type.  (object key casing being an example).